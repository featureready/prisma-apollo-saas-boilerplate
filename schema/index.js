import { gql } from 'apollo-server-express';

import userSchema from './user';
import teamSchema from './team';

const linkSchema = gql`
  scalar DateTime

  directive @requireAuth on FIELD_DEFINITION
  directive @analytics(type: String!, event: String) on FIELD_DEFINITION
  directive @analytics_group(type: String!, event: String) on FIELD_DEFINITION
  directive @computed(value: String) on FIELD_DEFINITION

  type Query {
    _: Boolean
  }
  type Mutation {
    _: Boolean
  }
  # type Subscription {
  #   _: Boolean
  # }
`;

export default [linkSchema, userSchema, teamSchema];
