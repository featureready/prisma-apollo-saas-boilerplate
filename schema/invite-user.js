import { gql } from 'apollo-server-express';

export default gql`
  extend type Mutation {
    InviteUser(input: InviteUserInput!): Invite! @requireAuth
  }

  input InviteUserInput {
    firstName: String!
    lastName: String!
    email: String!
  }

  type Invite {
    id: String!
    email: String!
    firstName: String!
    lastName: String!
    status: String!
  }
`;
