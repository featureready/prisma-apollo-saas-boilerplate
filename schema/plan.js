import { gql } from 'apollo-server-express';

export default gql`
  extend type Query {
    Plans: [Plan] @requireAuth
  }

  type Plan {
    id: String!
    title: String
    description: String
    amount: Int!
    currency: String!
    interval: String!
  }
`;
