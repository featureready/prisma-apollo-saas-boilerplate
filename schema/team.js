import { gql } from 'apollo-server-express';

export default gql`
  extend type Mutation {
    createTeam(name: String!): Team!
      @requireAuth
      @analytics(type: "track", event: "Create Team")
      @analytics_group(type: "group")
    updateTeam(name: String!): Team! @requireAuth @analytics(type: "track", event: "Update Team")
  }

  type Team {
    id: String
    name: String
  }
`;
