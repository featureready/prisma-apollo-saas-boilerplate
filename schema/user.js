import { gql } from 'apollo-server-express';

export default gql`
  extend type Query {
    me: User @requireAuth
    paymentHistory: [Invoice] @requireAuth
  }

  extend type Mutation {
    signup(firstName: String!, lastName: String!, email: String!, password: String!): AuthPayload!
      @analytics(type: "identify")
    login(email: String!, password: String!): AuthPayload! @analytics(type: "identify")
    forgotPassword(email: String!): Result
    resetPassword(password: String!, token: String!): Result
    updateUser(firstName: String, lastName: String, email: String, password: String): User!
      @requireAuth
    deleteUser(id: ID!): Int!
    addCreditCard(token: String): Result @requireAuth
    subscribePlan(planId: String!): Result @requireAuth
  }

  type AuthPayload {
    jwt: String!
    user: User!
  }

  type Result {
    message: String
  }

  type User {
    id: String
    firstName: String
    lastName: String
    fullName: String @computed(value: "$firstName $lastName")
    email: String
    createdAt: DateTime
    updatedAt: DateTime
    periodStart: DateTime
    periodEnd: DateTime
    team: Team
  }

  type Invoice {
    amountDue: Int
    amountPaid: Int
    invoicePdf: String
    status: String
    date: DateTime
    periodStart: DateTime
    periodEnd: DateTime
  }
`;
