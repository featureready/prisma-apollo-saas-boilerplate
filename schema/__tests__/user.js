import { mockServer } from 'graphql-tools';
import schema from '../index';
import {
  ME,
  PAYMENT_HISTORY,
  SIGNUP,
  LOGIN,
  UPDATE_USER,
  FORGOT_PASSWORD,
  DELETE_USER,
  ADD_CREDIT_CARD,
  SUBSCRIBE_PLAN
} from '../../queries/user';

const mocks = {
  Boolean: () => false,
  ID: () => '1',
  Int: () => 1,
  Float: () => 12.34,
  String: () => 'String',
  DateTime: () => new Date(2018, 8, 8)
};

const server = mockServer(schema, mocks, false);

describe('User schema', () => {
  it('has valid type definitions', () => {
    expect(() => server.query(`{ __schema { types { name } } }`)).not.toThrow();
  });

  it(`query { me } should return expected value`, () => {
    const variables = {};
    expect(server.query(ME, variables)).resolves.toMatchSnapshot();
  });

  it(`query { paymentHistory } should return expected value`, () => {
    const variables = {};
    expect(server.query(PAYMENT_HISTORY, variables)).resolves.toMatchSnapshot();
  });

  it(`mutation { signUp } should return expected value`, () => {
    const variables = {
      firstName: 'John',
      lastName: 'Kueh',
      email: 'john@beaconmaker.com',
      password: 'test123'
    };

    expect(server.query(SIGNUP, variables)).resolves.toMatchSnapshot();
  });

  it(`mutation { login } should return expected value`, () => {
    const variables = {
      email: 'john@beaconmaker.com',
      password: 'test123'
    };

    expect(server.query(LOGIN, variables)).resolves.toMatchSnapshot();
  });

  it(`mutation { forgotPassword } should return expected value`, () => {
    const variables = {
      email: 'john@beaconmaker.com'
    };

    expect(server.query(FORGOT_PASSWORD, variables)).resolves.toMatchSnapshot();
  });

  it(`mutation { updateUser } should return expected value`, () => {
    const variables = {
      firstName: 'John',
      lastName: 'Kueh',
      email: 'john@beaconmaker.com',
      password: 'test123'
    };

    expect(server.query(UPDATE_USER, variables)).resolves.toMatchSnapshot();
  });

  it(`mutation { deleteUser } should return expected value`, () => {
    const variables = {
      id: 1
    };

    expect(server.query(DELETE_USER, variables)).resolves.toMatchSnapshot();
  });

  it(`mutation { addCreditCard } should return expected value`, () => {
    const variables = {
      token: 'tok_1234'
    };

    expect(server.query(ADD_CREDIT_CARD, variables)).resolves.toMatchSnapshot();
  });

  it(`mutation { subscribePlan } should return expected value`, () => {
    const variables = {
      planId: 'plan_123'
    };

    expect(server.query(SUBSCRIBE_PLAN, variables)).resolves.toMatchSnapshot();
  });
});
