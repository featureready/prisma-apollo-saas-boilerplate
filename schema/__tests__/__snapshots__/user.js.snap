// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User schema mutation { addCreditCard } should return expected value 1`] = `
Object {
  "data": Object {
    "addCreditCard": Object {
      "message": "String",
    },
  },
}
`;

exports[`User schema mutation { deleteUser } should return expected value 1`] = `
Object {
  "data": Object {
    "deleteUser": 1,
  },
}
`;

exports[`User schema mutation { forgotPassword } should return expected value 1`] = `
Object {
  "data": Object {
    "forgotPassword": Object {
      "message": "String",
    },
  },
}
`;

exports[`User schema mutation { login } should return expected value 1`] = `
Object {
  "data": Object {
    "login": Object {
      "jwt": "String",
      "user": Object {
        "email": "String",
        "firstName": "String",
        "id": "String",
        "lastName": "String",
      },
    },
  },
}
`;

exports[`User schema mutation { signUp } should return expected value 1`] = `
Object {
  "data": Object {
    "signup": Object {
      "jwt": "String",
      "user": Object {
        "email": "String",
        "firstName": "String",
        "id": "String",
        "lastName": "String",
      },
    },
  },
}
`;

exports[`User schema mutation { subscribePlan } should return expected value 1`] = `
Object {
  "data": Object {
    "subscribePlan": Object {
      "message": "String",
    },
  },
}
`;

exports[`User schema mutation { updateUser } should return expected value 1`] = `
Object {
  "data": Object {
    "updateUser": Object {
      "email": "String",
      "firstName": "String",
      "id": "String",
      "lastName": "String",
    },
  },
}
`;

exports[`User schema query { me } should return expected value 1`] = `
Object {
  "data": Object {
    "me": Object {
      "createdAt": 2018-09-07T14:00:00.000Z,
      "email": "String",
      "firstName": "String",
      "fullName": "String",
      "id": "String",
      "lastName": "String",
      "periodEnd": 2018-09-07T14:00:00.000Z,
      "periodStart": 2018-09-07T14:00:00.000Z,
      "team": Object {
        "id": "String",
        "name": "String",
      },
      "updatedAt": 2018-09-07T14:00:00.000Z,
    },
  },
}
`;

exports[`User schema query { paymentHistory } should return expected value 1`] = `
Object {
  "data": Object {
    "paymentHistory": Array [
      Object {
        "amountDue": 1,
        "amountPaid": 1,
        "date": 2018-09-07T14:00:00.000Z,
        "invoicePdf": "String",
        "periodEnd": 2018-09-07T14:00:00.000Z,
        "periodStart": 2018-09-07T14:00:00.000Z,
        "status": "String",
      },
      Object {
        "amountDue": 1,
        "amountPaid": 1,
        "date": 2018-09-07T14:00:00.000Z,
        "invoicePdf": "String",
        "periodEnd": 2018-09-07T14:00:00.000Z,
        "periodStart": 2018-09-07T14:00:00.000Z,
        "status": "String",
      },
    ],
  },
}
`;
